test/restart/test_uda.py

```diff
@@ -0,0 +1,110 @@
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum_with_reduce}(pk) FROM {table}")]
+    assert res == [[4950]]
+
+    # Create a UDA that concatenates strings from all rows
+    concat = unique_name()
+    manager.cql.execute(f"CREATE FUNCTION {keyspace}.{concat} (x text, y text) RETURNS NULL ON NULL INPUT RETURNS text LANGUAGE lua AS 'return x..y'")
+
+    return_input_string = unique_name()
+    manager.cql.execute(f"CREATE FUNCTION {keyspace}.{return_input_string} (acc text) RETURNS NULL ON NULL INPUT RETURNS text LANGUAGE lua AS 'return acc'")
+
+    manager.cql.execute(f"CREATE AGGREGATE {sum} (text) SFUNC {concat} STYPE text FINALFUNC {return_input_string} INITCOND ''")
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum}(tx) FROM {table} WHERE pk IN (1,3,5)")]
+    assert res == [['aaaaaaaaa']]
+
+    # Create a variant of the UDA that uses a REDUCEFUNC
+    manager.cql.execute(f"CREATE AGGREGATE {sum_with_reduce} (text) SFUNC {concat} STYPE text REDUCEFUNC {concat} FINALFUNC {return_input_string} INITCOND ''")
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum_with_reduce}(tx) FROM {table} WHERE pk IN (1,3,5)")]
+    assert res == [['aaaaaaaaa']]
+
+    # Restart server before re-using AGGREGATE
+    await manager.server_restart(servers[0])
+    await manager.driver_connect()
+
+    # Check that all UDFs and UDAs still work after restart
+    res = [list(row) for row in manager.cql.execute(f"SELECT {keyspace}.{add}(pk, pk) FROM {table} WHERE pk = 1")]
+    assert res == [[2]]
+    res = [list(row) for row in manager.cql.execute(f"SELECT {keyspace}.{concat}(tx, tx) FROM {table} WHERE pk = 1")]
+    assert res == [['aa']]
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum}(pk) FROM {table}")]
+    assert res == [[4950]]
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum_with_reduce}(pk) FROM {table}")]
+    assert res == [[4950]]
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum}(tx) FROM {table} WHERE pk IN (1,3,5)")]
+    assert res == [['aaaaaaaaa']]
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum_with_reduce}(tx) FROM {table} WHERE pk IN (1,3,5)")]
+    assert res == [['aaaaaaaaa']]
+
+    # Check that we cannot create new UDAs with the same signatures, because they already exist
+    with pytest.raises(InvalidRequest, match="already exists"):
+        manager.cql.execute(f"CREATE AGGREGATE {sum}(int) SFUNC {add} STYPE int")
+    with pytest.raises(InvalidRequest, match="already exists"):
+        manager.cql.execute(f"CREATE AGGREGATE {sum}(text) SFUNC {concat} STYPE text")
+    with pytest.raises(InvalidRequest, match="already exists"):
+        manager.cql.execute(f"CREATE AGGREGATE {sum_with_reduce}(int) SFUNC {add} STYPE int")
+    with pytest.raises(InvalidRequest, match="already exists"):
+        manager.cql.execute(f"CREATE AGGREGATE {sum_with_reduce}(text) SFUNC {concat} STYPE text")
+
+    # Drop everything
```

Comment:
If this test fails somewhere, it leave the unique keyspace and everything in it behind.
But I guess it's fine in this test suite - because it can't be run on a pre-existing cluster anyway (unlike test/cql-pytest and test/alternator which can run against an existing long-term service and should clean up after themselves), and in a temporary cluster we don't care if things stay behind.

Reply:
Right, when using dtests we also have a similar case where we only have a temporary cluster